python/lsst/daf/butler/registry/dimensions/table.py

```diff
@@ -158,13 +228,560 @@ def sync(self, record: DimensionRecord) -> bool:
+            # them all up for one insert.
+            for gv, group in grouped.items():
+                overlapRecords.extend(self._compute(group, skypix[gv], gv))
+            _LOG.debug(
+                "Inserting %d new skypix overlap rows for %s where %s in %s.",
+                len(overlapRecords), self.element.name, self._governor.element.name, grouped.keys()
+            )
+            self._db.insert(self._overlapTable, *overlapRecords)
+
+    def _compute(
+        self,
+        records: Sequence[DimensionRecord],
+        skypix: NamedKeyDict[SkyPixSystem, List[int]],
+        governorValue: str,
+    ) -> Iterator[dict]:
+        """Compute all overlap rows for a particular governor dimension value
+        and all of the skypix dimensions for which its overlaps are enabled.
+
+        This method should only be called by `insert`.
+
+        Parameters
+        ----------
+        records : `Sequence` [ `DimensionRecord` ]
+            Records for ``self.element``.  Records with `None` regions are
+            ignored.  All must have the governor value given.
+        skypix : `NamedKeyDict` [ `SkyPixSystem`, `list` [ `int` ] ]
+            Mapping containing all skypix systems and levels for which overlaps
+            should be computed, grouped by `SkyPixSystem`.
+        governorValue : `str`
+            Value of this element's governor dimension for which overlaps
+            should be computed.  For example, if ``self.element`` is ``visit``,
+            this is an instrument name; if ``self.element`` is ``patch``, this
+            is a skymap name.
+
+        Yields
+        ------
+        row : `dict`
+            Dictionary representing an overlap row.
+        """
+        # Process input records one at time, computing all skypix indices for
+        # each.
+        for record in records:
+            if record.region is None:  # type: ignore
+                continue
+            assert getattr(record, self._governor.element.name) == governorValue
+            for system, levels in skypix.items():
+                if not levels:
+                    continue
+                baseOverlapRecord = record.dataId.byName()
+                baseOverlapRecord["skypix_system"] = system.name
+                levels.sort(reverse=True)
+                # Start with the first level, which is the finest-grained one.
+                # Compute skypix envelope indices directly for that.
+                indices: Dict[int, Set[int]] = {levels[0]: set()}
+                for begin, end in system[levels[0]].pixelization.envelope(record.region):  # type: ignore
+                    indices[levels[0]].update(range(begin, end))
+                # Divide those indices by powers of 4 (and remove duplicates)
+                # work our way up to the last (coarsest) level.
+                for lastLevel, nextLevel in zip(levels[:-1], levels[1:]):
+                    factor = 4**(lastLevel - nextLevel)
```

Comment:
Is this true for all known pixelization systems? Or do we only care about HTM?

Reply:
I am fairly certain it is true for HTM, Q3C, and healpix (for the "nested" healpix numbering system that we'd use, at least).